### ISSUE TITLE
Rule proposal: `prefer-change-array-by-copy`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1514]
["Change Array by copy" proposal](https://github.com/tc39/proposal-change-array-by-copy#change-array-by-copy)

## Fail

```js
array.reverse();
```

```js
array.sort();
```

```js
const array = [...foo];
array.splice(0, 1);
```

```js
const array = [...foo];
array[1] = 'changed';
```

## Pass

```js
array = array.toReversed();
```

```js
array = array.toSorted();
```

```js
const array = foo.toSpliced(0, 1);
```

```js
const array = foo.with(1, 'changed');
```

**Comment by sindresorhus** [comment]
Good idea. This will have to wait until it's shipping in a browser/Node.js version. The naming of the methods are not final.

**Comment by fregante** [comment]
I don't know, if the intent is to replace the existing variable, I don't particularly see an advantage in using with*

However this should be an error:

```JavaScript 
const newArray = old.reverse()
```

**Comment by fisker** [comment]
The proposal has changed see https://github.com/tc39/proposal-change-array-by-copy/commit/752e576db3d55e9d1d80a97df1f1e54c26ffbd7f

**Comment by fisker** [comment]
[Change Array by Copy to Stage 3 per 2022.03.30 TC39](https://github.com/tc39/proposals/pull/406#)

**Comment by fregante** [comment]
One pattern that I've been using is to clone arrays in the function parameters and I hope it can be preserved. I think the same thing applies to array destructuring.

#### Before

```js
function foo(originalBar) {
	const bar = [...originalBar];
	return bar;
}
```
#### After

```js
function foo([...bar]) {
	return bar;
}
```

**Comment by Cherry** [comment]
This [hit Stage 4 today](https://github.com/tc39/proposals/commit/ad4df8435f27f39eda26db3b940ae151980c8015), so we should see it start shipping soon more than likely 👀

**Comment by fisker** [comment]
Maybe we should split this into separate rules, to make rules easier to maintain.

- `prefer-array-to-reversed`
- `prefer-array-to-sorted`
- `prefer-array-to-spliced`
- `prefer-array-with`

**Comment by joealden** [comment]
With the release of Firefox 115, these methods are now available in all major browsers!

**Comment by klimashkin** [comment]
85%+ support according to caniuse

**Comment by silverwind** [comment]
Should also detect cases where `Array.from` is used to shallow-clone, e.g. when [`unicorn/prefer-spread`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-spread.md) is disabled:

### Fail

```js
const sorted = Array.from(arr).sort();
```

### Pass

```js
const sorted = arr.toSorted();
```

**Comment by markcarroll** [comment]
Bump. @sindresorhus this has been around for a few months in all browsers now and in Node 20, so any chance we could get this in now?

**Comment by sindresorhus** [comment]
This is accepted. Pull requests are welcome :)

**Comment by zbeyens** [comment]
Here is a gist that works for me, with spread and `Array.from` support https://gist.github.com/zbeyens/d4a2e70a7bf0687dffc6fbf7265d4ce5

**Comment by zanminkian** [comment]
Since `no-array-sort` and `no-array-reverse` have landed, `no-array-splice` is needed.
